source/common/router/router.cc

```diff
@@ -778,5 +763,37 @@ ProdFilter::createRetryState(const RetryPolicy& policy, Http::HeaderMap& request
                                 priority);
 }
 
+void Filter::UpstreamRequest::setRequestEncoder(Http::StreamEncoder& request_encoder) {
+  request_encoder_ = &request_encoder;
+  // Now that there is an encoder, have the connection manager inform the manager when the
+  // downstream buffers are overrun.  This may result in immediate watermark callbacks referencing
+  // the encoder.
+  parent_.callbacks_->addDownstreamWatermarkCallbacks(downstream_watermark_manager_);
+}
+
+void Filter::UpstreamRequest::clearRequestEncoder() {
+  // Before clearing the encoder, unsubscribe from callbacks.  This may result
+  // in some clean up calls to onBelowWriteBufferLowWatermark which reference
```

Comment:
Hm, what if unsubscribe took an argument for unwinding?    We could call unsubscribe(unwind=true) onRequestComplete() and unsubscribe(unwind=false) in all other cases.

H2 will sort itself out in the codec if a stream is simply reset so the only case we really have to worry about is HTTP reused connections which should have onRequestComplete, right?